cirq/optimizers/stratify.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Type, Callable, Union, Iterable
+
+from cirq import ops, circuits
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def stratified_circuit(
+        circuit: 'cirq.Circuit',
+        *,
+        categories: Iterable[Union[
+            'cirq.Gate',
+            'cirq.Operation',
+            Type['cirq.Gate'],
+            Type['cirq.Operation'],
+            Callable[['cirq.Operation'], bool]
```

Comment:
Optional: I think it'd be useful to also allow single callable that returns a category (e.g. as `int`). This would allow you to pass in a simple thing like `lambda g: len(g.qubits)` instead of an iterable of callables.

This is surprisingly tricky.

Reply:
Ah, a `key=` parameter. That's an interesting idea. It would be a pretty big change to the implementation.